datafusion/sql/src/unparser/plan.rs

```diff
@@ -455,19 +457,43 @@ impl Unparser<'_> {
             }
             LogicalPlan::SubqueryAlias(plan_alias) => {
                 let (plan, mut columns) =
-                    subquery_alias_inner_query_and_columns(&plan_alias);
+                    subquery_alias_inner_query_and_columns(plan_alias);
 
-                if !columns.is_empty() && !self.dialect.supports_column_alias_in_table_alias() {
-                    // if columns are returned than plan corresponds to a projection
-                    let LogicalPlan::Projection(inner_p) = plan else {
-                        return plan_err!(
-                            "Inner projection for subquery alias is expected"
-                        );
+                if !columns.is_empty()
+                    && !self.dialect.supports_column_alias_in_table_alias()
+                {
+                    // Instead of specifying column aliases as part of the outer table, inject them directly into the inner projection
+                    let plan_rewrite_result = match plan {
+                        LogicalPlan::Projection(inner_p) => {
+                            Ok(inject_column_aliases(inner_p, columns))
+                        }
+                        _ => {
+                            // projection is wrapped by other operator (LIMIT, SORT, etc), iterate through the plan to find it
+                            plan.to_owned()
+                                .map_children(|child| {
+                                    if let LogicalPlan::Projection(p) = &child {
+                                        Ok(Transformed::yes(inject_column_aliases(
+                                            p,
+                                            columns.clone(),
+                                        )))
+                                    } else {
+                                        Ok(Transformed::no(child))
+                                    }
+                                })
+                                .map(|plan| plan.data)
+                        }
+                    };
+
+                    let rewritten_plan = match plan_rewrite_result {
+                        Ok(p) => p,
+                        Err(e) => {
+                            return internal_err!(
+                                "Failed to transform SubqueryAlias plan: {e}"
+                            )
+                        }
                     };
 
-                    // Instead of specifying column aliases as part of the outer table inject them directly into the inner projection
-                    let rewritten_plan = inject_column_aliases(&inner_p, &columns);
-                    columns.clear();
+                    columns = vec![];
```

Comment:
```suggestion
                    columns.clear()
```

can we still use `.clear()`?

Reply:
Yeah, columns.clear() can't be used as columns are consumed/moved during call below

```
LogicalPlan::Projection(inner_p) => {
      Ok(inject_column_aliases(inner_p, columns))
  }
```